Refactor printers and describers into their own package.

smarterclayton opened the pull request:
This sets the stage for using printer code from the server side (decoupled from kubectl) and loosens the coupling between kubectl and the printers. `pkg/printers` contains interfaces and has an import restriction against pulling in API specific code, while `pkg/printers/internalversion` can be used for internal types.

Add a method on `Factory` for retrieving PrinterForCommand which uses the Scheme and RESTMapper from the Factory, not the hardcoded ones.  This further separates kubectl from the core API scheme and allows better composition.

Change NamePrinter to use RESTMapper (previously it was hardcoding those conversions). This means that we now return plural resource names (`pods/foo`) but is correct once aliases and shortnames start being returned by the mapper.

This is a prerequisite for server side get, but is pure refactor (contains no new features).

@deads2k @liggitt

smarterclayton committed:
Move, but do not refactor, printers to their own package

smarterclayton committed:
Refactor printers to separate typed printers from generic code

smarterclayton committed:
NamePrinter should not hardcode scheme

smarterclayton committed:
command Factory should provide Printers

The factory knows all possible types, and should abstract the process of
creating all printers. A future refactor can further reduce the
dependencies between printer code and internal types.

smarterclayton committed:
Restrict imports for printers

smarterclayton committed:
Refactor commands to use new factory method

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41729)
<!-- Reviewable:end -->

smarterclayton commented:
@kubernetes/sig-cli-pr-reviews @kubernetes/sig-api-machinery-pr-reviews since both sigs are impacted.

smarterclayton commented:
First two commits are mechanical separation of the code into the generic package or the typed package.  The others are cleanup and remove various todos leading to that split.

smarterclayton committed:
generated: bazel

smarterclayton commented:
@k8s-bot unit test this

Kind of shady, not sure if this is a brand new flake on `pkg/controller/podautoscaler` or not

deads2k commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -37,7 +37,7 @@ import (
 	tokenutil "k8s.io/kubernetes/cmd/kubeadm/app/util/token"
 	bootstrapapi "k8s.io/kubernetes/pkg/bootstrap/api"
-	"k8s.io/kubernetes/pkg/kubectl"
+	"k8s.io/kubernetes/pkg/printers"
```
Why top level and not under `kubectl`?  I'm assuming its because you anticipate them moving into the apiserver in a couple releases?

smarterclayton replied:
They'll be used from both for a while, then gradually moved only to apiserver.

fabianofranz replied:
Also we have the desire to eventually move `kubectl` out of the main repo, so I'd say top level makes sense.

deads2k commented on pkg/kubectl/cmd/util/factory_client_access.go:
```diff
@@ -392,8 +394,10 @@ func (f *ring0Factory) SuggestedPodTemplateResources() []schema.GroupResource {
-	return kubectl.NewHumanReadablePrinter(options), nil
+func (f *ring0Factory) Printer(mapping *meta.RESTMapping, options printers.PrintOptions) (printers.ResourcePrinter, error) {
+	p := printers.NewHumanReadablePrinter(options)
+	printersinternal.AddHandlers(p)
```
I'm reading top to bottom, so maybe this becomes clear, but I'm really surprised to find some of registration happening inside a factory method that could be called multiple times.

smarterclayton replied:
It registers it into the new object that it returns.

deads2k commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -151,15 +156,22 @@ func (f *ring1Factory) UnstructuredClientForMapping(mapping *meta.RESTMapping) (
 
-func (f *ring1Factory) Describer(mapping *meta.RESTMapping) (kubectl.Describer, error) {
+func (f *ring1Factory) PrinterForCommand(cmd *cobra.Command) (printers.ResourcePrinter, bool, error) {
+	mapper, typer := f.Object()
```
No peer references.  This will break downstream in openshift and anyone else trying to use it.  You'll need to move this to ring2.

deads2k commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -369,8 +381,8 @@ func (f *ring1Factory) AttachablePodForObject(object runtime.Object) (*api.Pod,
-func (f *ring1Factory) PrinterForMapping(cmd *cobra.Command, mapping *meta.RESTMapping, withNamespace bool) (kubectl.ResourcePrinter, error) {
-	printer, generic, err := PrinterForCommand(cmd)
+func (f *ring1Factory) PrinterForMapping(cmd *cobra.Command, mapping *meta.RESTMapping, withNamespace bool) (printers.ResourcePrinter, error) {
+	printer, generic, err := f.PrinterForCommand(cmd)
```
No peer references.  You'll need to push this out further.  Also, composition models suck compared to inheritance.

smarterclayton replied:
In this case it's just code reuse.  You'd have to override both commands.  I mainly was trying to eliminate the special case method, but I can bring it back.

smarterclayton replied:
I moved the common code back out into a private method so it's clear it's purely reuse.

deads2k commented on pkg/printers/customcolumn.go:
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 */
 
-package kubectl
+package printers
```
are all these just the world's most boring move?

smarterclayton replied:
First two columns are.

smarterclayton replied:
Commits

deads2k commented:
Some issues in the factory.  The move look ok, but we'll have to keep slicing later.

smarterclayton committed:
UnstructuredList should return 'items' set to the children

The set of items is not mutable (can't add or remove items) but the list
now is. Needs to be improved to make mutability is clear.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -73,11 +73,23 @@ func (obj *Unstructured) UnstructuredContent() map[string]interface{} {
+		for i, item := range obj.Items {
+			items[i] = item.Object
+		}
+		out["items"] = items
```
should always set "items", even if only to an empty list

smarterclayton replied:
done

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -73,11 +73,23 @@ func (obj *Unstructured) UnstructuredContent() map[string]interface{} {
+
+// UnstructuredContent returns a map contain an overlay of the Items field onto
+// the Object field. Items always overwrites overlay. Changing "items" in the
+// returned object will not affect the map.
```
the comment is misleading... changing "items" in the returned object does affect `obj.Object`, right? and the objects in the returned `items` list are also not deep copies

smarterclayton replied:
done, added a TODO

fabianofranz commented on pkg/printers/humanreadable.go:
```diff
@@ -0,0 +1,159 @@
+// more elegant output. It is not threadsafe, but you may call PrintObj repeatedly; headers
+// will only be printed if the object type changes. This makes it useful for printing items
+// received from watches.
+type HumanReadablePrinter struct {
```
Ha, I had a branch proposing this move to a separate file. It was growing huge. :)

fabianofranz commented on pkg/printers/common.go:
```diff
@@ -0,0 +1,41 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
2017?

smarterclayton replied:
This file was a move

fabianofranz commented on pkg/printers/.import-restrictions:
```diff
@@ -0,0 +1,10 @@
+{
+  "Rules": [
+    {
+      "SelectorRegexp": "k8s[.]io/kubernetes/pkg/(api$|apis/)",
```
Sanity check, you're matching `pkg/api` and `pkg/apis/<foo>`, but not the root  `pkg/apis`, correct?

smarterclayton replied:
`pkg/apis` doesn't have anything in it so i ignored it.

smarterclayton commented:
@k8s-bot gci gke e2e test this
@k8s-bot cvm gke e2e test this

smarterclayton commented:
@k8s-bot gci gke e2e test this

smarterclayton commented:
@k8s-bot gci gce e2e test this

smarterclayton commented:
@k8s-bot cvm gce e2e test this

smarterclayton commented:
@k8s-bot cvm gke e2e test this

smarterclayton commented:
Green, like a second look.

fabianofranz commented:
@k8s-bot gce etcd3 e2e test this

fabianofranz commented:
LGTM and most welcome, @deads2k @liggitt anything else?

deads2k commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -143,15 +148,26 @@ func (f *ring1Factory) UnstructuredClientForMapping(mapping *meta.RESTMapping) (
+}
+
+func (f *ring1Factory) PrinterForCommand(cmd *cobra.Command) (printers.ResourcePrinter, bool, error) {
+	return printerForRing1Factory(f, cmd)
```
@smarterclayton @fabianofranz  I think this still breaks because you're accessing a peer method. If someone tries to override `ring1Factory.Object()` and then expects to be able to delegate for this method, then the wrong `f` will be used and the wrong `f.Object()` will be called.

smarterclayton replied:
Addressed by moving to ring2

smarterclayton commented:
@k8s-bot non-cri e2e test this

smarterclayton commented:
@k8s-bot cvm gce e2e test this

smarterclayton commented:
@k8s-bot verify test this

smarterclayton commented:
@k8s-bot cvm gce e2e test this

smarterclayton commented:
@k8s-bot cvm gce e2e test this

liggitt commented:
LGTM

smarterclayton commented:
@k8s-bot verify test this
@k8s-bot bazel test this
@k8s-bot gce gce e2e test this

fabianofranz commented:
@k8s-bot gci gce e2e test this
